transforms/ExplicitConsdes.cpp

```diff
@@ -0,0 +1,189 @@
+                entry.function, "");
+            call->setDebugLoc(DebugLoc::get(0,0, before->getDebugLoc().getScope()));
+            if (after) {
+                call->insertAfter(after);
+            } else {
+                call->insertBefore(before);
+            }
+            after = call;
+        }
+    }
+
+   public:
+    static char ID;
+
+    ExplicitConsdes() : ModulePass(ID) {}
+
+    bool runOnModule(Module &mod) override {
+        using std::sort;
+        using std::tuple;
+        using std::vector;
+
+        bool change = false;
+
+        vector<FunctionEntry> ctors, dtors;
+
+        GlobalVariable *dtorsVar = nullptr;
+        GlobalVariable *ctorsVar = nullptr;
+
+        for (auto &glob : mod.globals()) {
+            if (glob.hasName() && glob.getName() == "llvm.global_dtors") {
+                fillFromInitializer(dtors, glob.getInitializer());
+                dtorsVar = &glob;
+            }
+
+            if (glob.hasName() && glob.getName() == "llvm.global_ctors" &&
+                glob.hasInitializer()) {
+                fillFromInitializer(ctors, glob.getInitializer());
+                ctorsVar = &glob;
+            }
+        }
+
+        if (ctors.empty() && dtors.empty()) {
+            return false;
+        }
+
+        // sort them to call order
+        sort(ctors.begin(), ctors.end(), priorityAsc);
+        sort(dtors.begin(), dtors.end(), priorityDesc);
+
+        // call constructors at the beginning of main
+        auto *main = mod.getFunction("main");
+        auto *ctorsBefore =
+            main->getEntryBlock().getFirstNonPHIOrDbgOrLifetime();
+        insertCalls(ctors, ctorsBefore);
+
+        vector<Instruction *> dtorsBefore;
+
+        // search for returns in main
+        for (auto &block : *main) {
+            if (isa<ReturnInst>(block.getTerminator())) {
+                dtorsBefore.push_back(block.getTerminator());
+            }
+        }
+
+        // and finally search for calls to exit in whole program
+        for (auto &func : mod.functions()) {
+            for (auto &block : func) {
+                for (auto &inst : block) {
+                    if (auto *call = dyn_cast<CallInst>(&inst)) {
+                        if (call->isIndirectCall()) continue;
+                        Function *called = call->getCalledFunction();
+                        if (!called) continue;
+                        auto *ty = called->getType()->getPointerElementType();
+                        if (called->hasName() && called->getName() == "exit" &&
+                            ty->getFunctionNumParams() == 1 &&
+                            ty->getFunctionParamType(0)->isIntegerTy()) {
+                            dtorsBefore.push_back(call);
+                        }
+                    }
+                }
+            }
+        }
+
+        for (auto *before : dtorsBefore) {
+            insertCalls(dtors, before);
+        }
+
+        if (ctorsVar) {
```

Comment:
I wonder whether some instruction may have a reference to the ctor/dtor global? In that case we need to drop also the references, so that we do not create a dangling reference.

Reply:
In LLVM yes, so I'll go ahead and fix it, but I can't think of any C program where this could happen.